Http Network Send Event Loop

thurstonsand opened the pull request:
just the stacked diff part of https://github.com/private-attribution/ipa/pull/221

The http implementation of the `Network` trait needs a simple event loop that takes data from the infra layer and sends it to the appropriate helper. This PR adds that functionality to initialization of the `HttpNetwork`

Commit:
send event loop

Commit:
use correct role in client

Commit:
Merge branch 'network-impl' into network-impl-send

martinthomson committed:
Be pedantic always

Though clippy::pedantic can backfire on occasion, this is easier to
manage.  We can deal with exceptions here as well.

Commit:
Upgrade to F_p

Commit:
Bitwise Sum protocol

Commit:
Add new test cases to test Fp32BitPrime

Commit:
Merge branch 'network-impl' into network-impl-send

Commit:
fixes to make it compile

Commit:
Merge branch 'receive-ordered-messages' into network-impl-send

Commit:
Add new test cases to test Fp32BitPrime

benjaminsavage committed:
Simplifying share conversion

benjaminsavage committed:
Merge DoubleRandom into ConvertShares

Commit:
increment offset for each chunk

Commit:
Merge remote-tracking branch 'origin/main' into network-impl-send

Commit:
Merge branch 'receive-ordered-messages' into network-impl-send

Commit:
remove whitespace

martinthomson committed:
clone_on_ref_ptr too

akoshelev committed:
Merge pull request #229 from private-attribution/pedantic

Be pedantic always

benjaminsavage committed:
Merge pull request #232 from private-attribution/simplify_share_conversion

Simplifying share conversion

benjaminsavage committed:
Merge pull request #233 from private-attribution/inline_double_random

Merge DoubleRandom into ConvertShares

benjaminsavage committed:
Merge pull request #231 from taikiy/bitwise_lt_test

New test cases to test BitwiseLessThan with Fp32BitPrime

akoshelev committed:
Add From<i32> implementation for RecordId (tests only)

I can't type `RecordId::from(0_u32)` anymore, sorry

Commit:
Address PR comments

martinthomson committed:
Stop making copies of Replicated

This doesn't help as much as I would have liked, and there are still
more calls to .clone() than would be ideal, but it does help, so I'm
doing it.

Lots of minor improvements to tests in here.

martinthomson committed:
Merge branch 'main' into replicated-no-copy

richajaindce committed:
[Minor optimization] GenBitPerm will now consume input vector rather than take a reference

richajaindce committed:
[Minor optimization] GenBitPerm will now consume input vector rather than take a reference

akoshelev committed:
Add sort bench

Very rudimentary, just to reproduce #237

the way you would run it (if you want traces):

```rust
RUST_LOG=trace cargo bench --bench oneshot_sort --profile dev --features test-fixture
```

akoshelev committed:
Fix formatting

akoshelev committed:
Merge pull request #238 from akoshelev/sort-bench

Add sort bench

richajaindce committed:
GenBitPerm can now work with both semi honest and malicious settings

Commit:
Random bit shares

martinthomson committed:
Inline some variables

martinthomson committed:
Merge branch 'main' into replicated-no-copy

martinthomson committed:
Rebase cleanup sort

Commit:
Merge branch 'receive-ordered-messages' into network-impl-send

martinthomson committed:
Merge pull request #235 from private-attribution/replicated-no-copy

Stop making copies of Replicated

akoshelev committed:
Fix reordering bug inside Gateway

Each time elements were inserted out of order to the send buffer inside the gateway, there was a risk of messages stuck inside the send buffer forever. This change addresses that and also adds a unit test that should prevent this from happening again.

akoshelev committed:
Merge from main

akoshelev committed:
Merge pull request #234 from akoshelev/test-i32-r

Add From<i32> implementation for RecordId (tests only)

richajaindce committed:
Passing input as a slice. Also, restructuring secureapplyinv and compose to have same template as bit_permutation.

richajaindce committed:
Rebase on main

richajaindce committed:
Passing share_of_one by value rather than reference since Sub/Add work on S - &S and, S + &S

richajaindce committed:
Cloning share_of_one

benjaminsavage committed:
Merge pull request #236 from richajaindce/bitperm_opt

[Minor optimization] GenBitPerm will now consume input vector rather …

benjaminsavage committed:
bye bye BoxError

martinthomson committed:
Merge pull request #242 from private-attribution/kill_those_boxes

bye bye BoxError

benjaminsavage committed:
make this test a bit better

akoshelev committed:
Remove deny(warnings)

This setting prevents compiler from building the code with warnings.
I think it is useful to be able to do that because often I prototype
an implementation and want to see whether it compiles and it is ok
if there are unadressed warnings such asunused imports.
Also I want to be able to run tests and get feedback from them without
addressing all the warnings.

I think we have good protection at PR level as GH action
would prevent pushing code that violates this no warnings rule.

benjaminsavage committed:
PR comments

benjaminsavage committed:
Merge pull request #244 from private-attribution/improve_specialized_mul_tests

make this test a bit better

benjaminsavage committed:
Minor change to the tests

benjaminsavage committed:
Merge pull request #245 from private-attribution/minor_tests_change

Minor change to the tests

Commit:
Rebase

benjaminsavage committed:
Merge pull request #219 from taikiy/carries

Bitwise Carries protocol

richajaindce committed:
Restructuring generate_sort_permutation also

Commit:
Address PR comments

benjaminsavage committed:
Merge pull request #220 from taikiy/bitadd

Bitwise Sum protocol

richajaindce committed:
Test got lost while merging with main

benjaminsavage committed:
Merge pull request #239 from richajaindce/malicious_bitperm

Gen bit perm for Semi honest and malicious setting

benjaminsavage committed:
very small fix

benjaminsavage committed:
Merge pull request #249 from private-attribution/small_nit

very small fix

benjaminsavage committed:
Make share_malicious accept r

Commit:
Fix [r]_b -> [r]_p computation logic

akoshelev committed:
Add -D warnings back to the pre commit hook

akoshelev committed:
Remove merge artifacts

akoshelev committed:
Merge pull request #246 from akoshelev/deny-deny-warnings

Remove deny(warnings)

akoshelev committed:
Run lint/Clippy first before build in GH actions

Most of the time, PR is broken because of formatting and/or Clippy issues. Therefore, it is desirable to run these checks first and get the feedback earlier before running builds and tests. My observation (likely biased towards my own behavior) is that we rarely submit code that does not compile or tests that are broken.

akoshelev committed:
Merge pull request #250 from private-attribution/share_malicious

Make share_malicious accept r

Commit:
Merge branch 'receive-ordered-messages' into network-impl-send

Commit:
improve test ergonomics

martinthomson committed:
Merge pull request #251 from akoshelev/lint-first

Run lint/Clippy first before build in GH actions

martinthomson committed:
Make FSV a little leaner

This doesn't need to hold the batch size, it can be told that as an
argument to take().

It doesn't need to return what was previously there when inserting
values: that is a bug and we can assert if it happens.

Rename some variables and functions to use shorter names.

(I'm still not a fan of this taking relative offsets: it makes the
encapsulation incomplete.  That can wait though.)

martinthomson committed:
Add a TODO

martinthomson committed:
Merge pull request #252 from private-attribution/simpler-fsv

Simpler FSV

akoshelev committed:
Context trait

Need to clean up the following

* Remove `bind`
* Relax `narrow`
* Introduce a small context struct that clone will refer to

akoshelev committed:
Remove `bind` and checks from `narrow`

What is left:

* Introduce a small context struct that clone will refer to

akoshelev committed:
ContextInner

akoshelev committed:
Formatting and Clippy

akoshelev committed:
Create folder for protocol contexts

akoshelev committed:
Bring back step string validation in debug mode

akoshelev committed:
Get rid of Cow inside protocol contexts

akoshelev committed:
Delete share_of_one module

akoshelev committed:
Remove commented lines (can someone create a clippy rule for that?)

Commit:
Merge remote-tracking branch 'fork/receive-ordered-messages' into network-impl-send

akoshelev commented on src/net/http_network.rs:
```diff
@@ -18,74 +24,78 @@ use tokio_stream::wrappers::ReceiverStream;
+                    .send_messages(args)
+                    .await
+                    .unwrap_or_else(|err| {
+                        tracing::error!("could not send message to client: {err}");
```
do you really want to swallow error here? If you fail to send, it is very unlikely that helpers can continue computation, but for some time system will pretend that everything is fine. If you need to build resiliency here, you could try to plug in backchannel to every message that you need to send (oneshot channel works here) and signal through it that something went wrong on the send side. Then, you'll leave it up to the sender to decide what they want to do - abort or retry.

Having said that, I don't think we need it right now and I would suggest to just panic here

thurstonsand replied:
yeah, you're right. i will just panic.

Commit:
just panic on client failure

richajaindce committed:
Changing BoxError to Error in sort protocols

richajaindce committed:
Introducing Reshare trait for SecretSharing and upgrading reshare for malicious

richajaindce committed:
Introducing Reshare trait for SecretSharing and upgrading reshare for malicious

martinthomson committed:
Merge pull request #258 from richajaindce/remove_boxerrors

Changing BoxError to Error in sort protocols

richajaindce committed:
Minor fix

richajaindce committed:
Minor fix

martinthomson committed:
Improve test ergonomics: semi-honest

This is a simple wrapper for test logic that might help improve how
tests run.

Commit:
Coerce a secret number is less than the prime

martinthomson committed:
Wrapper for joining 3 things

The futures crate has one of these, but it returns a tuple.  That is
more flexible than this, which isn't strictly what we need.

(Not sure if we need this with other active changes inbound, but I wrote
it and am putting it up anyway.)

akoshelev committed:
Merge pull request #262 from martinthomson/join3

Wrapper for joining 3 things

akoshelev committed:
Remove GAT from Reveal

and make it look the same as SecureMul. We don't need this advanced
feature anymore

martinthomson committed:
Use an array for PRSS in tests

Commit:
Bit-Decomposition Protocol

akoshelev committed:
s/ProtocolContext/Context

akoshelev committed:
Fix formatting

richajaindce committed:
[Sort Optimization] ApplyInv and Compose use the same random permutation in shuffle. The parties receive random permutation and invokes shuffle with reveal. Thus, we can commonalize them: we can shuffle and reveal sort permutation once and share with applyinv and compose.
Regarding security, this reuse doesn't affect it since we are still uniformly randomly permuting the permutation vector.

This is OPTAPPLYINV and OPTCOMPOSE from the paper

Commit:
Generate the smallest number of bits possible to form the prime number

Commit:
Rebase

akoshelev committed:
Specialize ContextInner for SemiHonest/Malicious impls

benjaminsavage committed:
Merge pull request #240 from taikiy/random_bits

Random Bits protocol

akoshelev committed:
Merge from main

benjaminsavage committed:
Merge pull request #264 from richajaindce/optshufflesreveal

[Sort Optimization] ApplyInv and Compose to use the same random permutation

martinthomson committed:
Merge pull request #263 from martinthomson/prss-array

Use an array for PRSS in tests

richajaindce committed:
Minor fixes from #264

richajaindce committed:
Minor fixes from #264

akoshelev committed:
Merge from main

akoshelev committed:
Merge pull request #225 from thurstonsand/receive-ordered-messages

Enforce ordered messages on server

martinthomson committed:
Merge pull request #253 from akoshelev/protocol-context-a-trait

Protocol context is a trait

martinthomson committed:
Inject RNG; rebase

akoshelev committed:
Merge pull request #261 from martinthomson/simplify-test

Simplify test

richajaindce committed:
Rebase on main

richajaindce committed:
Rebase on main

richajaindce committed:
Introducing validate_and_reconstruct_malicious test_fixture

richajaindce committed:
Introducing validate_and_reconstruct_malicious test_fixture

richajaindce committed:
Renaming tests to semi_honest and malicious

richajaindce committed:
Renaming tests to semi_honest and malicious

richajaindce committed:
Sending random permutation as a slice and not vector whenever possible

richajaindce committed:
Sending random permutation as a slice and not vector whenever possible

richajaindce committed:
Renaming generate_sort_permutation --> generate_permutation

richajaindce committed:
Renaming generate_sort_permutation --> generate_permutation

richajaindce committed:
Replace try_join_all by join3

richajaindce committed:
Replace try_join_all by join3

richajaindce committed:
Shuffle upgrade

richajaindce committed:
secureapplyinv upgrade

richajaindce committed:
compose upgrade

richajaindce committed:
Reveal_permutation upgrade

richajaindce committed:
sort upgrade

Commit:
Merge main

akoshelev committed:
Refactor reshare tests

While working on #256, I noticed that reshare tests consistently fail concurrency test. It was caused by always using RecordId(0) in a loop that terminates if reshare successfully generates unique shares. When loop needs more than 1 iteration it fails on PRSS uniqueness check.

While it was possible to fix that test, I did not really like the probabilistic nature of it and decided to create 2 tests instead. One validates that reshare protocol actually communicates and another one that it is correct.

It was also a good opportunity to use new test API implemented by @martinthomson in #261. It should be used everywhere.

One small change I made to the `TestWorld` to improve ergonomics changes `make_contexts` function to narrow generated contexts to some unique (within this world) step, so the following syntax is made possible

```rust
  for _ in 0..10 {
     world.semi_honest(..., |ctx, v| { ... })
  }

martinthomson committed:
A simpler malicious upgrade (WIP)

richajaindce committed:
removing explicit type declaration

akoshelev committed:
Merge pull request #257 from richajaindce/reshare_upgrade

Introducing Reshare trait for SecretSharing and upgrading reshare for malicious

akoshelev committed:
Merge from main

martinthomson committed:
anonymous lifetime is OK

Co-authored-by: Alex Koshelev <akoshelev@users.noreply.github.com>

martinthomson committed:
Merge pull request #268 from akoshelev/reshare-test-fix

Refactor reshare tests

martinthomson committed:
Compiling!

akoshelev committed:
Remove byteorder crate

Fewer dependencies is better, right?

martinthomson committed:
Merge pull request #270 from akoshelev/no-byteorder

Remove byteorder crate

martinthomson committed:
Add a sample malicious test and add some refinements

martinthomson committed:
Remove clippy override

martinthomson committed:
Rebase

martinthomson committed:
Merge branch 'main' into malicious-runner

martinthomson committed:
Simplify both bit_permutation tests equally

akoshelev committed:
Use PRSS seq rng to generate permutation

I am on a crusade to eliminate different sources of randomness in IPA
to keep PRSS and thread_rng() only. Having those prevents concurrency
tests to provide "replay" functionality.

I believe we shouldn't be using any PRNG as our PRSS implementation
already provides random generator that is suitable to shuffle vectors.

This change replaces ChaCha RNG with PRSS seq RNG.

martinthomson committed:
WIP

martinthomson committed:
Review feedback

martinthomson committed:
Merge pull request #269 from martinthomson/malicious-runner

A simpler malicious upgrade (WIP)

martinthomson committed:
WIP

richajaindce committed:
Merge branch 'main' into minor

richajaindce committed:
Merge branch 'minor' into upgrade_all

martinthomson committed:
Rework a whole bunch of tests

This includes support for validation of the x/rx malicious shares that
are produced by the protocol, which turned out to be annoying to
implement.

benjaminsavage committed:
Merge pull request #265 from richajaindce/minor

Minor fixes from #264 suggested by Martin

richajaindce committed:
Merge branch 'main' into upgrade_all

martinthomson committed:
Merge main

martinthomson committed:
Merge branch 'main' into validate-x-rx

martinthomson committed:
One less BoxError

benjaminsavage committed:
Merge pull request #267 from richajaindce/upgrade_all

Upgrade all subprotocols of sort

martinthomson committed:
Merge pull request #272 from martinthomson/validate-x-rx

Rework a bunch of tests

benjaminsavage committed:
Merge pull request #273 from martinthomson/less-boxerror

One less BoxError

Commit:
WIP

Commit:
Merge main

Commit:
RingBuffer is Mutex guarded

akoshelev committed:
Merge from main

akoshelev committed:
Merge pull request #271 from akoshelev/bye-bye-cha-cha

Use PRSS sequential randomness to generate permutation instead of ChaCha

Commit:
Place RandomBitsGenerator inside Context

Commit:
Address PR comments

benjaminsavage committed:
Implemeting a simple, straightforward bitwise les than protocol

benjaminsavage committed:
updated comments

benjaminsavage committed:
Adding a really dumb implementation of bitwise sum

benjaminsavage committed:
fix comment

benjaminsavage committed:
mt's comments

martinthomson committed:
Merge pull request #274 from private-attribution/dumb_bitwise_lt

Implemeting a simple, straightforward bitwise les than protocol

Commit:
Place RandomBitsGenerator inside ContextInner

Commit:
Merge remote-tracking branch 'origin/main' into bits

Commit:
Add align_bit_length() to dumb_bitwise_lt

benjaminsavage committed:
Merge pull request #266 from taikiy/bits

Bit-Decomposition protocol

benjaminsavage committed:
rebase

benjaminsavage committed:
Merge pull request #275 from private-attribution/dumb_bitwise_add

Dumb bitwise add

Commit:
Merge remote-tracking branch 'origin' into network-impl-send

thurstonsand commented:
closed because the base branch has been merged
